### ISSUE TITLE
check phone clock to avoid GPS issues (unix time, see mailing list)

### ISSUE THREAD

**Issue by Djfe** [mozilla/MozStumbler#1526]
We should warn the user if the phone clock is off (for example: Unix time: 1970, or if the phone clock timestamp is in general lower than the build timestamp) to avoid collecting wrong data

this is not only about wrong timestamps but also about the GPS
if the phone clock is off that far, then the GPS position is likely wrong as well, so we shouldn't collect data then and avoid uploading it.

Maybe we could even implement a time check (compare current phone to server time before uploading or something like that)

^this issue is a result of the following discussion on the mailing list:
https://groups.google.com/forum/#!topic/mozilla.dev.geolocation/biwiie__M10

**Comment by lua86** [comment]
Checking local unix time against build unix time should be both highly
effective and really efficient. Just that check should be enough.

**Comment by narfg** [comment]
I read the discussion on the mailing list but I'm not entirely convinced that an incorrect phone clock affects the GPS position. Can anyone provide a source for that?

**Comment by Djfe** [comment]
I thought the phone needs a very accurate time to get your current position
and according to http://www.maptoaster.com/maptoaster-topo-nz/articles/how-gps-works/how-gps-works.html it does need that indeed
but it doesn't use your phone clock
-> if you have at least 4 satellites your gps receiver will calculate an accurate time itself from the signals it receives

so an inaccurate phone clock might not be that bad

but I guess warning the user wouldn't do any harm and ichnaea would get correct timestamps for their uploads (recording date)

**Comment by narfg** [comment]
@Djfe Yes, the GPS receiver chip uses its own clock, not the one of the phone.
Does Mozilla Stumbler actually use the phone clock for its timestamps? The code in StumblerFilter.java seems to get the timestamp directly from the GPS module.

**Comment by hannosch** [comment]
The report we got was from someone using the wigle app and uploading data to us. So this might have actually been a bug in the wigle app. But there might also be some problems with how some phones get the GPS assistance data or internally confuse the GPS and phone times.
